File: Part 2 - Regression/Polynomial Regression/polinomial_regression.R
# Regresion Polinomica

#importando dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[,2:3]

#Dividir data set entre conjunto de entranamiento y testing
#library(caTools)
#set.seed(123) #Seleccionando semilla
#sample.split(Datos a separar, porcentaje para entrenar)
#split = sample.split(dataset$Purchased,SplitRatio = 0.8)
#Separando datos de entrenamiento y testing
#training_set = subset(dataset,split==TRUE)
#testing_set = subset(dataset,split==FALSE)

#Escalado de valores
#[filas,columnas]
#training_set[,2:3] = scale(training_set[,2:3])#[Todas las filas, de 2 a 3]
#testing_set[,2:3] = scale(testing_set[,2:3])

#Ajustar el modelo de regresion lineal simple con el conjunto de datos 
lin_reg = lm(formula = Salary ~ .,
               data = dataset)
summary(lin_reg)

#Ajustar el modelo de regresion polinomica con el conjunto de datos
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
dataset$Level5 = dataset$Level^5
poly_reg = lm(formula = Salary ~ .,
              data = dataset)
summary(poly_reg)

#Visualizacion del modelo lineal
library(ggplot2)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +#colocando puntos
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata=dataset)),
            colour = "blue") + #Colocando recta de regresion
  ggtitle("Prediccion lineal del sueldo en funcion del nivel del empleado")+
  xlab("Nivel del empleado") + 
  ylab("Sueldo (en $)")

#Visualizacion del modelo polinomico
library(ggplot2)
x_grid = seq(min(dataset$Level),max(dataset$Level),0.1)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +#colocando puntos
  geom_line(aes(x = x_grid, y = predict(poly_reg, newdata=data.frame(Level = x_grid,
                                                                     Level2 = x_grid^2,
                                                                     Level3 = x_grid^3,
                                                                     Level4 = x_grid^4,
                                                                     Level5 = x_grid^5))),
            colour = "blue") + #Colocando recta de regresion
  ggtitle("Prediccion polinomica del sueldo en funcion del nivel del empleado")+
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")


#Prediccion de nuevos resultado con Regrsion Lineal
y_pred = predict(lin_reg,newdata=data.frame(Level = 6.5))

#Prediccion de nuevos resultado con Regrsion Polinomica
y_pred = predict(poly_reg,newdata=data.frame(Level = 6.5, Level2 = 6.5^2, 
                                             Level3 =6.5^3, Level4 = 6.5^4,
                                             Level5 = 6.5^5))

File: Part 2 - Regression/Polynomial Regression/polinomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Mon Apr  3 16:13:59 2023

@author: josue
"""

# Regresion polinomica

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

dataset = pd.read_csv('Position_Salaries.csv')

# matriz. Obteniendo valores del data frame [filas,columnas]
X = dataset.iloc[:, 1:2].values #Matriz 
y = dataset.iloc[:, 2].values  # vector


#Dividir data set entre conjunto de entranamiento y testing

#X_train,X_test,y_train,y_test=train_test_split(X,y,test_size = 0.2, random_state=0)

# Ajustar la regresion lineal con el data set
from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Ajustar la regresion polinomica con el data set
from sklearn.preprocessing import PolynomialFeatures
#Se deben de agregar las x^n hacia la derecha 
poly_reg = PolynomialFeatures(degree = 10)#Por defecto n = 2
X_poly = poly_reg.fit_transform(X)#Salida: termino independiente + b1X1 + b2X1^2 + ... bnX1^n
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly,y)

# Visualización de los resultados del Modelo Lineal
plt.scatter(X, y, color = "red")
plt.plot(X, lin_reg.predict(X), color = "blue")
plt.title("Modelo de Regresión Lineal")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()

# Visualización de los resultados del Modelo Polinómico
X_grid = np.arange(min(X),max(X),0.1)
X_grid = X_grid.reshape(len(X_grid),1)
plt.scatter(X, y, color = "red")
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = "blue")
plt.title("Modelo de Regresión Polinomica")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()

# Prediccion de nuestros modelos
lin_reg.predict([[6.5]])
lin_reg_2.predict(poly_reg.fit_transform([[6.5]]))
